Better app bundle

ismagilli opened the pull request:
Note that PR marked as WIP (Work in Progress)

Related to #123

It would be great if someone besides me took part in the testing. Also, any suggestions and recommendations are welcome.


TODO list
- [x] Improve macOS integration my modifying source code (see `patch` directory)
  - [x] Connect to macOS specific "open file" signal
  - [x] Load plugins from bundle (`/path/to/.app/Contents/Resources/plugins`)
  - [x] Load plugins form bundle only if inside bundle (bundle has ID)
- [x] Add patch to zathura formula
- [x] Improve `convert-into-app` script
  - [x] Add list of extensions to `djvu`, `poppler` and `ps` extensions
  - [x] Add list of extensions to `cb` and `mupdf` extensions
  - [x] Improve `Info.plist`
    - [x] Add _about_ and _copyright_
    - [x] Add supported extensions
    - [x] Add zathura version
  - [x] Detect installed plugins
  - [x] Build list of supported extensions based on founded plugins
  - [x] Symlink installed plugins' `.dylib` into bundle
  - [x] `/opt/homebrew/bin/zathura` => `$(brew --prefix zathura)/bin/zathura`
  - [x] Fix typo in caveats: `zathura open example.pdf` => `zathura example.pdf`
  - [x] Updated icon URL to link to a file from the master branch, rather than a file from a specific commit
  - [x] Colorize output
- [x] Add info about creating bundles into README
- [x] Add info about other plugins into README

<details> 
<summary><b>Code from mac-integration.c from patch with color highlighting</b></summary>

```c
/* SPDX-License-Identifier: Zlib */

#include <girara/log.h>
#include <glib.h>
#include <gtkosxapplication.h>

#include "mac-integration.h"

#include "plugin.h"
#include "shortcuts.h"
#include "zathura.h"


static void load_plugins_from_bundle(zathura_t* zathura);
static gboolean openfile_callback(GtkosxApplication* app, gchar* path, zathura_t* zathura);


void zathura_mac_init(zathura_t* zathura) {
  load_plugins_from_bundle(zathura);

  GtkosxApplication* zathura_mac = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
  gtkosx_application_set_use_quartz_accelerators(zathura_mac, FALSE);
  gtkosx_application_ready(zathura_mac);

#if 0
  gchar* id = gtkosx_application_get_bundle_id();
  if (id != NULL) {
    girara_warning("TestIntegration Error! Bundle has ID %s", id);
  }
  g_free(id);
#endif

  g_signal_connect(G_OBJECT(zathura_mac), "NSApplicationOpenFile", G_CALLBACK(openfile_callback), zathura);
}

static void load_plugins_from_bundle(zathura_t* zathura) {
  zathura_plugin_manager_t* plugin_manager = zathura->plugins.manager;

  gchar* resource_path = gtkosx_application_get_resource_path();
  gchar* plugins_path = g_build_path("/", resource_path, "plugins", NULL);

  zathura_plugin_manager_set_dir(plugin_manager, plugins_path);
  zathura_plugin_manager_load(plugin_manager);

  g_free(resource_path);
  g_free(plugins_path);
}

static gboolean openfile_callback(GtkosxApplication* UNUSED(app), gchar* path, zathura_t* zathura) {
  if (zathura_has_document(zathura) == false) {
    document_open_idle(zathura, path, NULL, ZATHURA_PAGE_NUMBER_UNSPECIFIED, NULL, NULL, NULL, NULL);
    return true;
  }

  pid_t pid = fork();

  if (pid == 0) { /* child */
    if (setsid() == -1) {
      girara_error("Could not start new process group: %s", strerror(errno));
    }

    char *args[] = { gtkosx_application_get_executable_path(), path, NULL };
    execv(args[0], args);

    girara_error("Could not replaces process image: %s", strerror(errno));
    exit(1);
  } else if (pid < 0) { /* error */
    girara_error("Could not fork: %s", strerror(errno));
    /* Should we somehow notify user about this situation? */
  }

  return true;
}
```
</details>

SylvanFranklin commented:
You are a hero bro, if I'm rich and famous one day I'll remember how you helped out the open source zathura on mac community.

JayanAXHF commented:
Amazing work here! Thank you sooo much for your contribution here!

ismagilli commented:
First update. It is still work-in-progress.

I added TODO list into conversation message.
What do you think about some of the items from this todo list? Maybe some items should be removed completely, some should be moved to separate PRs? What can you say about text messages in a  script, in formulae, in a readme? Are there any recommendations or comments on the C code and script code?

Is it a good idea to copy (actually symlink) plugins' `.dylib` into bundle and load plugins from this directory on startup (inside C code)?

I have a bit strange idea. Add `brew --prefix zathura-$PLUGIN` for each into zathura lookup directory (it will make `brew` runtime dependency) What do you think?

Do we need this code?
```c
gchar* id = gtkosx_application_get_bundle_id();
if (id != NULL) {
  girara_warning("TestIntegration Error! Bundle has ID %s", id);
}
g_free(id);
```

SylvanFranklin commented:
Sorry I just got around to this, but thanks for all the work on it! Honestly I think you're more qualified to make the calls on symlinking and stuff than I am, but I'll look into it when I have time. I could make you a maintainer if you would like, I can tell you're invested in our success.

ismagilli commented:
WIP label has just been removed!

I was thinking about adding plugins to the bundle. It is better to avoid using brew in runtime. That's why I added symbolic links to plugins. And also added the loading of these plugins from inside the bundle.

This is not the best solution, as some plugins attempt to load them twice, which does not lead to problems, but displays a message in the console that the plugin was loaded earlier.

> I could make you a maintainer if you would like
It's nice to hear such a suggestion. But I'm not sure if my future contributions will be as active.

SylvanFranklin commented:
Wow dude you really did a lot for this project that's awesome. I'll merge it in later tonight! also would love to here @JayanAXHF's thoughts on changes, but this looks great to me!

JayanAXHF commented:
> Wow dude you really did a lot for this project that's awesome. I'll merge it in later tonight! also would love to here @JayanAXHF's thoughts on changes, but this looks great to me!

The changes look good to me. Thank you soo much for the contributions!
